### PR TITLE
[rpc] Correct reconsiderblock help text, add test

### PR DESCRIPTION
**Pull request by maflcko**
Rework documentation and test to match the implementation

### CODE REVIEW

**Comment by ch4ot1c** [comment]
utACK

**Comment by MarcoFalke** [comment]
For reference:

* https://drahtbot.github.io/reports/coverage/bitcoin/bitcoin/master/total.coverage/src/validation.cpp.gcov.html#2855
* https://drahtbot.github.io/reports/coverage/bitcoin/bitcoin/15057/total.coverage/src/validation.cpp.gcov.html#2855

**Comment by gmaxwell** [comment]
IIRC if you     have a chain  A B C D E and invalidate D then invalidate B, then reconsider  B you'll end back up on E.

**Comment by MarcoFalke** [comment]
Added a test for this as well (and fixed up the help text). Maybe the method should have been called `reconsiderchain` with the description "Reconsiders all chains the given block is in"

File: test/functional/rpc_invalidateblock.py [hunk 1]
```diff
@@ -21,46 +26,66 @@ def run_test(self):
...
+        wait_until(lambda: self.nodes[1].getblockcount() == 4, timeout=5)
+
+        self.log.info("Verify that we reconsider all ancestors as well")
+        blocks = self.nodes[1].generatetoaddress(10, ADDRESS_BCRT1_UNSPENDABLE)
+        assert_equal(self.nodes[1].getbestblockhash(), blocks[-1])
+        # Invalidate the two blocks at the tip
+        self.nodes[1].invalidateblock(blocks[-1])
+        self.nodes[1].invalidateblock(blocks[-2])
+        assert_equal(self.nodes[1].getbestblockhash(), blocks[-3])
+        # Reconsider only the previous tip
```
Comment (promag) [inline test/functional/rpc_invalidateblock.py]: s/previous/initial/

File: test/functional/rpc_invalidateblock.py [hunk 2]
```diff
@@ -21,46 +26,66 @@ def run_test(self):
...
-            raise AssertionError("Node 1 reorged to a lower height: %d"%node1height)
+        wait_until(lambda: self.nodes[2].getblockcount() == 3, timeout=5)
+        wait_until(lambda: self.nodes[0].getblockcount() == 4, timeout=5)
+        wait_until(lambda: self.nodes[1].getblockcount() == 4, timeout=5)
+
+        self.log.info("Verify that we reconsider all ancestors as well")
+        blocks = self.nodes[1].generatetoaddress(10, ADDRESS_BCRT1_UNSPENDABLE)
+        assert_equal(self.nodes[1].getbestblockhash(), blocks[-1])
+        # Invalidate the two blocks at the tip
+        self.nodes[1].invalidateblock(blocks[-1])
```
Comment (promag) [inline test/functional/rpc_invalidateblock.py]: remove?

File: test/functional/rpc_invalidateblock.py [hunk 3]
```diff
@@ -21,46 +26,66 @@ def run_test(self):
...
+        # Reconsider only the previous tip
+        self.nodes[1].reconsiderblock(blocks[-1])
+        # Should be back at the tip by now
+        assert_equal(self.nodes[1].getbestblockhash(), blocks[-1])
+
+        self.log.info("Verify that we reconsider all descendants")
+        blocks = self.nodes[1].generatetoaddress(10, ADDRESS_BCRT1_UNSPENDABLE)
+        assert_equal(self.nodes[1].getbestblockhash(), blocks[-1])
+        # Invalidate the two blocks at the tip
+        self.nodes[1].invalidateblock(blocks[-2])
```
Comment (promag) [inline test/functional/rpc_invalidateblock.py]: remove?

**Review by promag — COMMENTED** [review]
Some comments/questions, other than that looks great.

File: test/functional/rpc_invalidateblock.py (hunk 3 above)
Comment (MarcoFalke) [reply test/functional/rpc_invalidateblock.py]: Why? This is exactly what the test should be testing for. See https://github.com/bitcoin/bitcoin/pull/15057#issuecomment-450706123

Comment (promag) [reply test/functional/rpc_invalidateblock.py]: Yap sorry.

**Comment by promag** [comment]
utACK fa38d3d, don't mind the above nits..

Candidate for backport?

**Comment by MarcoFalke** [comment]
I'd say no. User-facing it is only a minor documentation fixup of a method that should not casually be used in production.

**Pull request merged**